core/src/main/kotlin/com/ritense/valtimo/importchangelog/ImportChangelogDeploymentService.kt

```diff
@@ -0,0 +1,121 @@
+    fun deployAll() {
+        logger.info { "Running import deployer" }
+        loadResources().forEach { resource ->
+            val filename = getFilename(resource)
+            val changesetDeployer = getChangesetDeployer(resource)
+            if (changesetDeployer != null) {
+                logger.info { "Running import deployer changelog: $filename" }
+                val resourceContent = resource.inputStream.bufferedReader().use { it.readText() }
+                deploy(changesetDeployer, resourceContent, filename)
+            }
+        }
+        logger.info { "Finished running import deployer" }
+    }
+
+    fun deploy(changesetDeployer: ChangesetDeployer, content: String, filename: String) {
+        val changesetDetails = changesetDeployer.getChangesetDetails(filename, content)
+        val md5sum = computeMd5sum(objectMapper.writeValueAsString(changesetDetails.valueToChecksum))
+        val existing = importChangesetRepository.findById(changesetDetails.changesetId).getOrNull()
+
+        if (existing != null) {
+            if (md5sum != existing.md5sum) {
+                throw RuntimeException("Computed checksum '$md5sum' doesn't match existing '${existing.md5sum}' for import $filename")
+            } else {
+                logger.debug { "Verified checksum '$md5sum' for import $filename" }
+            }
+        } else {
+            changesetDeployer.deploy(content)
+            saveChangeset(changesetDetails.changesetId, filename, md5sum)
+        }
+    }
+
+    private fun getFilename(resource: Resource) = resource.uri.toASCIIString().substringAfterLast("resources/")
+
+    private fun getChangesetDeployer(resource: Resource): ChangesetDeployer? {
+        val type = getTypeFromFileName(resource.filename)
+        return changesetDeployers.find { it.getType() == type }
+    }
+
+    private fun saveChangeset(changesetId: String, filename: String, md5sum: String) {
+        importChangesetRepository.save(
+            ImportChangeset(
+                id = changesetId,
+                filename = filename,
+                dateExecuted = Instant.now(),
+                orderExecuted = importChangesetRepository.getNextOrderExecuted() ?: 0,
+                md5sum = md5sum
+            )
+        )
+    }
+
+    private fun computeMd5sum(valueToChecksum: String): String {
+        return MD5Util.computeMD5(
+            Normalizer.normalize(
+                StringUtil.standardizeLineEndings(valueToChecksum)
+                    .replace("\uFFFD", ""), Normalizer.Form.NFC
+            )
+        )
+    }
+
+    private fun getTypeFromFileName(filename: String?): String? {
+        return if (filename == null) {
+            null
+        } else {
+            Regex(".+\\.(.+)\\.json").find(filename)?.groupValues?.get(1)
+        }
+    }
+
+    @Throws(IOException::class)
+    private fun loadResources() = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(PATH)
```

Comment:
Maybe pattern matching to actual resources should be done by the individual deployers.

Reply:
changed